File: frontend/src/engine/manifest.ts
/**
 * Resolved Manifest 读取层（docs/15 §四·补 / §十三 / §十五）。
 *
 * 服务端是唯一真相源：`availability` / `completion` / `artifacts` 一律读这里下发的值，
 * 前端**不得**重新推导「这个病例有没有某能力」「能不能结束训练」。
 *
 * 本模块是纯函数 + 类型：没有 React、没有网络、没有面板注册表。
 * renderer key → 组件 的映射见 `components/training/workspace/renderers.ts`。
 */

export const ACTIVITY_STATE_AVAILABLE = "available";

export const ARTIFACT_EMPTY = "empty";
export const ARTIFACT_DRAFT = "draft";
export const ARTIFACT_SUBMITTED = "submitted";

/** manifest.actions[].id：学生主动结束训练 */
export const ACTION_COMPLETE_SESSION = "complete_session";

/** completion.blockers[].code —— 前端只用于给 blocker 找到落点，不用于判定 */
export const CODE_SESSION_NOT_ACTIVE = "SESSION_NOT_ACTIVE";
export const CODE_ARTIFACT_NOT_SUBMITTED = "ARTIFACT_NOT_SUBMITTED";

export interface ManifestActivityAvailability {
	state: string;
	reason_code: string | null;
}

export interface ManifestActivityUI {
	/** 渲染组件 key（RendererMap 的键） */
	renderer: string;
	/** 形态：side_panel / …（当前只有侧栏/底部面板） */
	placement: string;
	/** 面板顺序（越小越靠前） */
	order: number;
}

export interface ManifestActivity {
	id: string;
	label: string;
	availability: ManifestActivityAvailability;
	commands: string[];
	ui: ManifestActivityUI;
	evidence_kind: string | null;
	artifact_kind: string | null;
}

export interface ManifestArtifact {
	required: boolean;
	state: string;
	submitted_at: string | null;
	updated_at: string | null;
}

export interface ManifestBlockerTarget {
	type: string;
	id: string;
}

export interface ManifestBlocker {
	code: string;
	message: string;
	target: ManifestBlockerTarget | null;
}

export interface ManifestCondition {
	id: string;
	label: string;
	satisfied: boolean;
}

export interface ManifestAction {
	id: string;
	label: string;
	enabled: boolean;
}

export interface SessionManifest {
	schema: string;
	projection: string;
	workflow: { id: string; label: string; ui: Record<string, string> };
	case: { case_id: number | null; revision_id: number | null; revision_no: number | null };
	session: { session_id: number | null; status: string; revision: number | null };
	activities: ManifestActivity[];
	artifacts: Record<string, ManifestArtifact>;
	completion: {
		eligible: boolean;
		conditions: ManifestCondition[];
		blockers: ManifestBlocker[];
	};
	actions: ManifestAction[];
}

// ── 窄化工具（服务端载荷是 JSON，逐个字段收敛，缺字段不猜）──────────────

function asRecord(value: unknown): Record<string, unknown> | null {
	return value !== null && typeof value === "object" && !Array.isArray(value)
		? (value as Record<string, unknown>)
		: null;
}

function text(value: unknown): string {
	return typeof value === "string" ? value : "";
}

function optionalText(value: unknown): string | null {
	return typeof value === "string" ? value : null;
}

function flag(value: unknown): boolean {
	return value === true;
}

function optionalNumber(value: unknown): number | null {
	return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function strList(value: unknown): string[] {
	return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
}

function parseActivity(raw: unknown): ManifestActivity | null {
	const item = asRecord(raw);
	if (!item) return null;
	const id = text(item.id);
	if (!id) return null;
	const availability = asRecord(item.availability) ?? {};
	const ui = asRecord(item.ui) ?? {};
	return {
		id,
		label: text(item.label) || id,
		availability: {
			state: text(availability.state),
			reason_code: optionalText(availability.reason_code),
		},
		commands: strList(item.commands),
		ui: {
			renderer: text(ui.renderer),
			placement: text(ui.placement) || "side_panel",
			order: optionalNumber(ui.order) ?? 0,
		},
		evidence_kind: optionalText(item.evidence_kind),
		artifact_kind: optionalText(item.artifact_kind),
	};
}

function parseArtifact(raw: unknown): ManifestArtifact | null {
	const item = asRecord(raw);
	if (!item) return null;
	return {
		required: flag(item.required),
		state: text(item.state) || ARTIFACT_EMPTY,
		submitted_at: optionalText(item.submitted_at),
		updated_at: optionalText(item.updated_at),
	};
}

function parseBlocker(raw: unknown): ManifestBlocker | null {
	const item = asRecord(raw);
	if (!item) return null;
	const target = asRecord(item.target);
	return {
		code: text(item.code),
		message: text(item.message),
		target: target ? { type: text(target.type), id: text(target.id) } : null,
	};
}

/**
 * 解析会话 manifest。
 *
 * 非 session 投影 / 结构不对时返回 `null` —— 调用方按「清单未到位」处理
 * （等待完整详情），绝不回退到前端自算的能力表。
 */
export function parseSessionManifest(raw: unknown): SessionManifest | null {
	const data = asRecord(raw);
	if (!data) return null;
	if (text(data.projection) !== "session") return null;
	const workflow = asRecord(data.workflow);
	const session = asRecord(data.session);
	if (!workflow || !session) return null;

	const activities = (Array.isArray(data.activities) ? data.activities : [])
		.map(parseActivity)
		.filter((activity): activity is ManifestActivity => activity !== null)
		.sort((a, b) => a.ui.order - b.ui.order);

	const artifacts: Record<string, ManifestArtifact> = {};
	for (const [kind, value] of Object.entries(asRecord(data.artifacts) ?? {})) {
		const artifact = parseArtifact(value);
		if (artifact) artifacts[kind] = artifact;
	}

	const completion = asRecord(data.completion) ?? {};
	const caseInfo = asRecord(data.case) ?? {};

	return {
		schema: text(data.schema),
		projection: text(data.projection),
		workflow: {
			id: text(workflow.id),
			label: text(workflow.label),
			ui: Object.fromEntries(
				Object.entries(asRecord(workflow.ui) ?? {}).map(([key, value]) => [key, text(value)]),
			),
		},
		case: {
			case_id: optionalNumber(caseInfo.case_id),
			revision_id: optionalNumber(caseInfo.revision_id),
			revision_no: optionalNumber(caseInfo.revision_no),
		},
		session: {
			session_id: optionalNumber(session.session_id),
			status: text(session.status),
			revision: optionalNumber(session.revision),
		},
		activities,
		artifacts,
		completion: {
			eligible: flag(completion.eligible),
			conditions: (Array.isArray(completion.conditions) ? completion.conditions : [])
				.map((raw) => {
					const item = asRecord(raw);
					if (!item) return null;
					return {
						id: text(item.id),
						label: text(item.label) || text(item.id),
						satisfied: flag(item.satisfied),
					};
				})
				.filter((condition): condition is ManifestCondition => condition !== null),
			blockers: (Array.isArray(completion.blockers) ? completion.blockers : [])
				.map(parseBlocker)
				.filter((blocker): blocker is ManifestBlocker => blocker !== null),
		},
		actions: (Array.isArray(data.actions) ? data.actions : [])
			.map((raw) => {
				const item = asRecord(raw);
				if (!item) return null;
				return { id: text(item.id), label: text(item.label) || text(item.id), enabled: flag(item.enabled) };
			})
			.filter((action): action is ManifestAction => action !== null),
	};
}

// ── 读取器（全部读服务端字段，不做任何推断）──────────────────────────────

/** 当前会话内可用的 Activity（顺序 = manifest 顺序 = 服务端 ui.order）。 */
export function availableActivities(manifest: SessionManifest | null): ManifestActivity[] {
	if (!manifest) return [];
	return manifest.activities.filter((activity) => activity.availability.state === ACTIVITY_STATE_AVAILABLE);
}

/**
 * blocker 对应的落点：`target.type === "artifact"` → 产出该产物的 Activity。
 *
 * 只回答「该跳进哪个面板」，不回答「能不能完成」——后者始终由服务端 eligible 决定。
 */
export function blockerActivity(manifest: SessionManifest | null, blocker: ManifestBlocker): ManifestActivity | undefined {
	if (blocker.target?.type !== "artifact") return undefined;
	return manifest?.activities.find((activity) => activity.artifact_kind === blocker.target?.id);
}

/**
 * 完成前置：workflow 声明必须已提交的产物 kinds（`manifest.artifacts[].required`）。
 *
 * 用于把 `/end` 的原子「提交并完成」载荷与服务端声明对齐，而不是前端判定能否结束。
 */
export function requiredArtifacts(manifest: SessionManifest | null): string[] {
	if (!manifest) return [];
	return Object.entries(manifest.artifacts)
		.filter(([, artifact]) => artifact.required)
		.map(([kind]) => kind);
}

export function completionBlockers(manifest: SessionManifest | null): ManifestBlocker[] {
	return manifest?.completion.blockers ?? [];
}

export function completionConditions(manifest: SessionManifest | null): ManifestCondition[] {
	return manifest?.completion.conditions ?? [];
}

File: backend/modules/training/manifest.py
"""Resolved Manifest —— 服务端解析出的单一真相源（docs/15 §四·补）。

同一 envelope 支持三种投影（session / catalog / authoring）；本切片只落地
``projection="session"``：学生运行时的 activities[].availability、artifacts
（draft/submitted）、completion（eligible/conditions/blockers）与 actions[].enabled。

硬规则：前端不得重新推导 availability / completion / audience / 病例版本 ——
``eligible`` 与 ``availability`` 一律来自这里。
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

from core.statuses import TrainingStatus
from modules.training.activities import artifact_definition, resolve_activities
from modules.training.profile import WorkflowDefinition

SCHEMA = "workflow-manifest"
PROJECTION_SESSION = "session"

ARTIFACT_EMPTY = "empty"
ARTIFACT_DRAFT = "draft"
ARTIFACT_SUBMITTED = "submitted"

#: 完成阻塞的机器可读码（前端只渲染 message 与 target，不自行判定）
CODE_SESSION_NOT_ACTIVE = "SESSION_NOT_ACTIVE"
CODE_ARTIFACT_NOT_SUBMITTED = "ARTIFACT_NOT_SUBMITTED"

ACTION_COMPLETE_SESSION = "complete_session"


@dataclass(frozen=True)
class ArtifactState:
    """学生产物状态（目前只有护理记录一种产物）。"""

    kind: str
    state: str = ARTIFACT_EMPTY
    required: bool = False
    submitted_at: str | None = None
    updated_at: str | None = None

    @property
    def submitted(self) -> bool:
        return self.state == ARTIFACT_SUBMITTED


def build_session_manifest(
    *,
    session_id: int,
    status: str,
    revision: int | None,
    case_data: Mapping[str, Any] | None,
    workflow: WorkflowDefinition,
    case_id: int | None = None,
    case_revision_id: int | None = None,
    case_revision_no: int | None = None,
    overrides: Mapping[str, Any] | None = None,
    artifacts: Mapping[str, ArtifactState] | None = None,
) -> dict[str, Any]:
    """解析会话 manifest（纯函数：入参齐备即可复算，无 IO）。"""
    session_active = status == TrainingStatus.IN_PROGRESS
    resolved_activities = resolve_activities(
        case_data,
        allowed=workflow.activities,
        overrides=overrides,
        session_active=session_active,
    )
    # 产物只在病例真的启用了对应 Activity 时才存在（与 finalize 的完成判定同一口径）
    configured_artifacts = {
        str(item.definition.artifact_kind)
        for item in resolve_activities(case_data, allowed=workflow.activities, overrides=overrides)
        if item.availability.available and item.definition.artifact_kind
    }
    resolved_artifacts = {
        kind: ArtifactState(
            kind=kind,
            state=artifact.state,
            required=workflow.completion.covers(kind),
            submitted_at=artifact.submitted_at,
            updated_at=artifact.updated_at,
        )
        for kind, artifact in (artifacts or {}).items()
        if kind in configured_artifacts
    }
    for kind in configured_artifacts:
        resolved_artifacts.setdefault(kind, ArtifactState(kind=kind, required=workflow.completion.covers(kind)))

    conditions: list[dict[str, Any]] = []
    blockers: list[dict[str, Any]] = []
    if not session_active:
        blockers.append(
            {
                "code": CODE_SESSION_NOT_ACTIVE,
                "message": "训练已结束，无法再次提交完成",
                "target": None,
            }
        )
    for kind in (k for k in workflow.completion.required_artifacts if k in configured_artifacts):
        artifact = resolved_artifacts.get(kind) or ArtifactState(kind=kind, required=True)
        definition = artifact_definition(kind)
        label = definition.label if definition else kind
        satisfied = artifact.submitted
        conditions.append({"id": f"{kind}_submitted", "label": f"提交{label}", "satisfied": satisfied})
        if session_active and not satisfied:
            blockers.append(
                {
                    "code": CODE_ARTIFACT_NOT_SUBMITTED,
                    "message": f"请先提交{label}，再结束训练",
                    "target": {"type": "artifact", "id": kind},
                }
            )

    eligible = session_active and all(condition["satisfied"] for condition in conditions)

    return {
        "schema": SCHEMA,
        "projection": PROJECTION_SESSION,
        "workflow": {"id": workflow.id, "label": workflow.label, "ui": dict(workflow.ui)},
        # 会话引用的病例版本（docs/15 §六）：旧记录（迁移前）没有 revision_id，
        # 只有 case_snapshot，因此两项都可能为 null。
        "case": {"case_id": case_id, "revision_id": case_revision_id, "revision_no": case_revision_no},
        "session": {"session_id": session_id, "status": status, "revision": revision},
        "activities": [
            {
                "id": resolved.definition.id,
                "label": resolved.definition.label,
                "availability": {
                    "state": resolved.availability.state,
                    "reason_code": resolved.availability.reason_code,
                },
                "commands": sorted(resolved.definition.commands),
                "ui": {
                    "renderer": resolved.definition.ui_renderer,
                    "placement": resolved.definition.ui_placement,
                    "order": resolved.definition.ui_order,
                },
                "evidence_kind": resolved.definition.evidence_kind,
                "artifact_kind": resolved.definition.artifact_kind,
            }
            for resolved in resolved_activities
        ],
        "artifacts": {
            kind: {
                "required": artifact.required,
                "state": artifact.state,
                "submitted_at": artifact.submitted_at,
                "updated_at": artifact.updated_at,
            }
            for kind, artifact in resolved_artifacts.items()
        },
        "completion": {"eligible": eligible, "conditions": conditions, "blockers": blockers},
        "actions": [
            {
                "id": ACTION_COMPLETE_SESSION,
                "label": "结束训练",
                "enabled": eligible,
            }
        ],
    }
